File: PadraoGof/Criacional/Singleton/Singleton01.py
#--------------------------------------------------------------------------------------------------
#Fatec Diadema - Luigi Papaiz 
#Curso : Desenvolvimento de Software Multiplataforma - DSM 
#Disciplina: Técnica de Programação II
#Nome: Kaiki Kenji       
#Data: 15/03/2024
#Descrição: Aplicação do padrão GOF - Criacional - Singleton
#Exemplo do Código do Livro de Padrão de Projetos - GOF
#--------------------------------------------------------------------------------------------------

class Singleton:
    instance = None

    def __new__(cls):
        if cls.instance is None:
            cls.instance = super().__new__(cls)
            cls.instance.value = 0
        return cls.instance
    
    def increment(self):
        self.value += 1
        print(f"Value: {self.value}")

#Cliente
        
s1 = Singleton()
s2 = Singleton()

print (s1 == s2)

s1.increment()
s2.increment()

File: PadraoGof/Criacional/Singleton/Singleton01.js
/* 
--------------------------------------------------------------------------------------------------
Fatec Diadema - Luigi Papaiz 
Curso : Desenvolvimento de Software Multiplataforma - DSM 
Disciplina: Técnica de Programação II
Nome: Kaiki Kenji       
Data: 15/03/2024
Descrição: Aplicação do padrão GOF - Criacional - Singleton
Exemplo do Código do Livro de Padrão de Projetos - GOF
--------------------------------------------------------------------------------------------------
*/

class Singleton {
    constructor() {
        if (Singleton.instance == null) {
            Singleton.instance = this;
            this.value = 0;
        }
        return Singleton.instance;
    }

    increment() {
        this.value += 1;
        console.log(`Value: ${this.value}`);
    }
}

//Cliente

const s1 = new Singleton();
const s2 = new Singleton();

console.log(s1 === s2);

s1.increment();
s2.increment();